### ISSUE TITLE
!strcmp(locale, "C"):Error:Assert failed:in file baseapi.cpp, line 209

### ISSUE THREAD

**Issue by charhuang** [sirfz/tesserocr#165]
import tesserocr
from PIL import Image

image = Image.open('image.png')
print(tesserocr.image_to_text(image))

Mac 10.14.1
tesserocr  2.3.1  
Python 3.6.7

**Comment by xiejun946** [comment]
i am the same error as you

**Comment by xiejun946** [comment]
try it as this ,i work well
import locale
locale.setlocale(locale.LC_ALL, 'C')
import tesserocr

**Comment by abhishek-27** [comment]
This can be fixed by either importing `export LC_ALL=C` in terminal or setting it in the python file where you are using tesserocr as `locale.setlocale(locale.LC_ALL, 'C')` or if still getting error in Pycharm IDE then just set LC_ALL=C as Environment variable in the run configuration for that file.

**Comment by jameshe3** [comment]
> try it as this ,i work well
> import locale
> locale.setlocale(locale.LC_ALL, 'C')
> import tesserocr

I met the same issue. Your solution helped . Thanks!

**Comment by bertsky** [comment]
It's mentioned in the [FAQ](https://github.com/tesseract-ocr/tesseract/wiki/4.0x-Common-Errors-and-Resolutions) as well. And BTW this measure is only meant to be [provisional](https://github.com/tesseract-ocr/tesseract/issues/1670#issuecomment-398027291), to be removed as soon as certain legacy parts of the code base (locale-dependent functions like `sscanf`) are gone.

You might have problems with this workaround though: other parts of your code might depend localization, e.g. on `*.UTF-8`. In that case, you can still try to set the locale to `C` merely temporarily (just before  importing from tesserocr), and maybe also set back afterwards (but the latter will risk giving sub-optimal results or even crash during recognition).

**Comment by nijel** [comment]
This is really ridiculous, it makes it nearly impossible to use the module in a program using locales. Does anybody have idea what kind of breakages can man expect if locales are changed after importing the module? Or is the recommended approach to set C locale before every tesserocr call and restore correct locales after it?

PS: I've at least tried to leverage the limitation to support utf-8 locale as well in https://github.com/tesseract-ocr/tesseract/pull/2272 as without that Python 3 before 3.7 has problems handling unicode filenames...

**Comment by bertsky** [comment]
Indeed. I'll have a chance to ask @stweil about it tomorrow. The utf-8 patch was a good idea, thanks!

What you dug up about locale switching in Python does make matters worse. So maybe this should really be done on the C level in tesserocr? Has anyone tried that?

Interestingly, [gImageReader](https://github.com/manisandro/gImageReader/blob/0f05810d30fad471a1cf0949f711f437c3120dfc/gtk/src/Recognizer.cc#L151) does temporarily reset the locale, but only for Tesseract initialisation, not during recognition. And both [libopencv-contrib/modules/text](https://github.com/opencv/opencv_contrib/tree/master/modules/text) and [ffmpeg/libavfilter](https://github.com/FFmpeg/FFmpeg/blob/97f47fd601d0e380109f501e4aabff71ffbad052/libavfilter/vf_ocr.c#L55) do nothing about it (but seem to be based on version 3 still).

**Comment by stweil** [comment]
There is a whole number of functions in the standard C library which depend on the current locale, not only `sscanf`, but also `sprintf` and their relatives and several classifier functions. See my comment https://github.com/tesseract-ocr/tesseract/issues/1670#issuecomment-399528335.

This causes problems like https://github.com/tesseract-ocr/tesseract/issues/1250.

With my German locale de_DE.UTF-8, for example, Tesseract would read wrong float and double values from existing configuration files because German writes such numbers like `3,14159` instead of `3.14159`. A file written with that locale would be incompatible with most of the rest of the world.

https://github.com/tesseract-ocr/tesseract/commit/db9c7e0312cbfc7b71d4ca968bc0ac48273d676c shows how that can be fixed. Similar fixes are still needed for the rest of the Tesseract code. Then the hard requirement of a fixed locale setting and the assertions can be removed.

**Comment by nijel** [comment]
I understand the problems (I've fixed similar issues in many programs). I just find strange that `import tesserocr` can lead to terminating Python. That's not really what I would expect from any library.

Also it would be great to know where all the C locale is needed to make things work properly. As @ bertsky pointed out there are already programs switching locales for tesseract initialization, will that work properly or should it be there for the recognition as well (as outlined in https://github.com/tesseract-ocr/tesseract/issues/1670#issuecomment-431560537)? Having this documented would make it easier to implement this properly.

**Comment by stweil** [comment]
Switching locale just to avoid the assertion and switching it back after Tesseract initialization will leave you with the risk of problems caused by a "wrong" locale. Depending on your own locale and on the code parts which you use it can work. If your locale does not use the comma as a decimal point and your primary characters are based on Latin, chances are high that it will work. With comma or with an Asian locale, chances are much lower. So to summarize, a temporary locale switch is not a general solution usable for everybody.

**Comment by stweil** [comment]
> Also it would be great to know where all the C locale is needed to make things work properly.

Just search the code for the problematic function names. It is some work to decide when those code locations are used (text recognition or training, old recognizer or LSTM, ...). Therefore I think it is easier to replace all those code parts by code which does not depend on locale settings.

**Comment by nijel** [comment]
With the locales switching I'm trying to figure out solution that would work now. It's certainly not nice, but I really don't see other way to use tesserocr.

The major problem is that upcoming Debian stable will come with tesseract 4.0.0, so apparently my code will have to deal with that version for *years*, no matter what you fix in upcoming tesseract releases. Not sure if something can be done in the tesserocr module to mitigate this or at least to make possible some runtime detection avoiding terminating Python...

**Comment by mikkelee** [comment]
I'm currently doing this in Python3.7:

```python
@contextmanager
def c_locale():
    try:
        currlocale = locale.getlocale()
    except ValueError:
        currlocale = ('en_US', 'UTF-8')
    log.debug(f'Switching to C from {currlocale}')
    locale.setlocale(locale.LC_ALL, "C")
    yield
    log.debug(f'Switching to {currlocale} from C')
    locale.setlocale(locale.LC_ALL, currlocale)

def get_hocr(image):
    with c_locale():
        from tesserocr import PyTessBaseAPI
        with PyTessBaseAPI() as tesseract:
            tesseract.SetImage(image)
            hocr = tesseract.GetHOCRText(0)
    return hocr
```

It appears to work, as all access to Tesseract is done inside a C-locale.

**Comment by nijel** [comment]
Similar workaround has been posted in https://github.com/tesseract-ocr/tesseract/issues/1670#issuecomment-431560537, it's pretty straightforward to implement, however looking at Python docs makes me a bit worried if that is good thing to do (see https://github.com/sirfz/tesserocr/issues/165#issuecomment-467607251). Anyway I've implemented this in https://github.com/WeblateOrg/weblate/commit/6724204977e7c466a463595736cc5299c8f01829 as I don't see better way to address this for now.

**Comment by agnelvishal** [comment]
After typing `export LC_ALL=C` in the terminal, run the python code in the same terminal. Running the python code in different terminal window won't work.

**Comment by stweil** [comment]
Tesseract 4.1 and Tesseract 5 no longer require such locale hacks.

**Comment by mikkelee** [comment]
I can't find anything about Tesseract 5?

**Comment by stweil** [comment]
Tesseract 5 will be the next release. Tesseract Git master is the latest version targeting 5.0, [5.0.0-alpha](https://github.com/tesseract-ocr/tesseract/tree/5.0.0-alpha) is a pre-release.

**Comment by mikkelee** [comment]
Ah, thanks

**Comment by sirfz** [comment]
Solved in Tesseract v4.1
